plugins/Serial/__init__.py

```diff
@@ -111,14 +72,42 @@ def MyHexDecoder(input):
 ]
 
 
+def RevokeWorkPermit(self):
+    if self.plugin.ReceiveThread is not None:
+        if self.plugin.ReceiveThread_WorkPermit.is_set() is True:
+            self.plugin.ReceiveThread_WorkPermit.clear()
+            time.sleep(self.plugin.SerialPortInterface.timeout)
```

Comment:
the use of time.sleep should never be used to stall a thread loop. The reason why is there is no exit in the event EG gets shutdown. the thread will simply get stopped and the file for the port will not get closed properly. Can cause issues.

you already have a threading.Event item right there use that instead of the time.sleep. a threading.Event() instance will bump out of the wait if the event gets set.

    if self.plugin.ReceiveThread is not None:
        if self.plugin.ReceiveThread_WorkPermit.is_set():
            self.plugin.ReceiveThread_WorkPermit.clear()
            self.plugin.ReceiveThread_WorkPermit.wait(self.plugin.SerialPortInterface.timeout)

Reply:
This sleep is not for stalling the thread, but to wait until the current loop of the thread is processed to avoid errors (the threading.Event() only blocks a thread at certain points in the code, where "wait" is defined).

Reply:
I do not remember off hand how long EG waits for threads to terminate I would have to go and check. But I do know that if there is no exit strategy from a thread being temporarily stalled EG will gripe about it in the debugging log. The main reason is because of things not getting closed properly and EG simply just shutting down, It can lead to orphaned sockets where only a reboot can fix the issue. or a lock not getting removed on a com port, and the fix is the same as the socket. a reboot would have to take place. It's just good practice to always provide a mechanism that will exit a pause in a thread.

here is an example. read the commented lines it will tell you how to use the example. you will understand what I am talking about.

    import threading
    
    event = threading.Event()
    
    def do():
        print 'thread starting'
        while not event.isSet():
            print 'thread looping'
    
            # run the script once.
            # comment and uncomment the lines below as labeled
            # run the script again
    
            # comment me
            time.sleep(10.0)
    
            # uncomment me
            # event.wait(10.0)
    
        print 'thread exiting'
    
    t = threading.Thread(target=do)
    t.start()
    
    import time
    event.wait(1.0)
    
    print 'setting event'
    event.set()
    print 'script should exit now'
    
    start = time.time()
    t_count = 0
    
    while t.isAlive():
        if t_count == 0:
            if time.time() - start >= 1:
                t_count += 1
                print 'script waiting to exit:', t_count, 'seconds'
        if time.time() - start - t_count >= 1:
            t_count +=1
            print 'script waiting to exit:', t_count, 'seconds'
    
    print 'script finally exiting'